src/cmds/pbs_rsub.c

```diff
@@ -696,8 +767,131 @@ main(int argc, char *argv[], char *envp[])
 		}
 	}
 
+	if (is_maintenance_resv) {
+		int i;
+		char tmp_str[BUF_SIZE];
+
+		pbs_errno = 0;
+		bstat_head = pbs_statvnode(connect, "", NULL, NULL);
+		if (bstat_head == NULL) {
+			if (pbs_errno) {
+				errmsg = pbs_geterrmsg(connect);
+				if (errmsg != NULL) {
+					fprintf(stderr, "pbs_rsub: %s\n", errmsg);
+				} else {
+					fprintf(stderr, "pbs_rsub: Error (%d) submitting reservation\n", pbs_errno);
+				}
+			} else {
+				fprintf(stderr, "pbs_rsub: No nodes found\n");
+			}
+
+			CS_close_app();
+			exit(pbs_errno);
+		}
+
+		/* here, the select is crafted from hosts */
+		hostp = maintenance_hosts;
+		for (; *hostp; hostp++) {
+			/* lets craft select_str */
+			int host_ncpus = 0;
+
+			for (bstat = bstat_head; bstat; bstat = bstat->next) {
+				for (pattr = bstat->attribs; pattr; pattr = pattr->next) {
+					if (pattr->resource && strcmp(pattr->name, ATTR_rescavail) == 0 && strcmp(pattr->resource, "host") == 0)
+						host = pattr->value;
+					if (pattr->resource && strcmp(pattr->name, ATTR_rescavail) == 0 && strcmp(pattr->resource, "ncpus") == 0)
+						ncpus = pattr->value;
+				}
+
+				if (strcmp(host, *hostp) == 0 && atoi(ncpus) > 0) {
+					host_ncpus += atoi(ncpus);
+				}
+			}
+
+			if (host_ncpus > 0) {
+				if (!select_str) {
+					select_str = malloc(BUF_SIZE);
+					if (select_str == NULL) {
+						fprintf(stderr, "pbs_rsub: Out of memory\n");
+						exit(2);
+					}
+					select_str_size = BUF_SIZE;
+
+					sprintf(select_str, "select=host=%s:ncpus=%d", *hostp, host_ncpus);
+				} else {
+					sprintf(tmp_str, "+host=%s:ncpus=%d", *hostp, host_ncpus);
+
+					if (pbs_strcat(&select_str, &select_str_size, tmp_str) == NULL) {
+						fprintf(stderr, "pbs_rsub: Out of memory\n");
+						exit(2);
+					}
+				}
+			}
+		}
+
+		/* here, the execvnodes is crafted from vnodes */
+		hostp = maintenance_hosts;
+		for (; *hostp; hostp++) {
```

Comment:
@prakashcv13 If I understand correctly, you have suggested combining the crafting of 'select' and 'execvnodes' into one loop. It is possible but IMHO it will harm the readability pretty badly, and it will not improve the complexity. I would prefer keeping the crafting of 'select' and 'execvnodes' separately.

Reply:
@vchlum - I would vouch for performance instead of readability. What do you think?

Reply:
This change is not included in the last commit yet. Let me know if you really prefer to combine the two for loops into one.

Reply:
@prakashcv13 Well, I think the performance gain will be quite small, but no problem. OK, I will change it into one for loop in next commit. I will add some comments if necessary.

Reply:
Done